### PR TITLE
[JSC] Only load/call synthetic @instanceFieldInitializer if needed

### PR DESCRIPTION
**Pull request by caitp**
Adds metadata to track whether or not a class constructor (or arrow/eval in a
class constructor) needs to initialize instance fields, and only emits
the property load/call if it is needed. This also removes the test to
see if the method is undefined.

The cost is slightly more metadata to be collected, including extra
fields in various places (like ExecutableInfo,
UnlinkedFunctionExecutable, EvalExecutable, etc), and checking the
entire class elements list to see if there are any instance fields.
Sould be fairly minor, and will probably avoid regression we saw for
classes with no instance fields in the original implementation.

### CODE REVIEW

File: Source/JavaScriptCore/llint/LLIntCommon.h [hunk 1]
```diff
@@ -29,7 +29,7 @@
 // - Prints every instruction executed if Options::traceLLIntExecution() is enabled.
 // - Prints some information for some of the more subtle slow paths if
 //   Options::traceLLIntSlowPath() is enabled.
-#define LLINT_TRACING 0
+#define LLINT_TRACING 1
```
Comment (caitp) [inline Source/JavaScriptCore/llint/LLIntCommon.h]: oops!

File: Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp [hunk 2]
```diff
@@ -2831,15 +2831,16 @@ RegisterID* BytecodeGenerator::emitCreateThis(RegisterID* dst)
 
 RegisterID* BytecodeGenerator::emitInstanceFieldInitializationIfNeeded(RegisterID* dst, RegisterID* constructor, const JSTextPosition& divot, const JSTextPosition& divotStart, const JSTextPosition& divotEnd)
 {
-    Ref<Label> done = newLabel();
+    if (!(isConstructor() || isDerivedConstructorContext()) || !classContextHasInstanceFields())
+        return dst;
```
Comment (caitp) [inline Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: I guess this should probably be an assert instead

File: Source/JavaScriptCore/builtins/BuiltinExecutableCreator.cpp [hunk 3]
```diff
@@ -32,7 +32,7 @@ namespace JSC {
 
 UnlinkedFunctionExecutable* createBuiltinExecutable(VM& vm, const SourceCode& source, const Identifier& ident, ConstructorKind kind, ConstructAbility ability)
 {
-    return BuiltinExecutables::createExecutable(vm, source, ident, kind, ability);
+    return BuiltinExecutables::createExecutable(vm, source, ident, kind, ability, false);
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/BuiltinExecutableCreator.cpp]: I think we should use an enumerator instead. This improves a lot readability. Something like ```ClassFieldsInitializer::Needed``` and ```ClassFieldsInitializer::NotNeeded```

File: Source/JavaScriptCore/builtins/BuiltinExecutables.cpp [hunk 4]
```diff
@@ -58,30 +58,30 @@ SourceCode BuiltinExecutables::defaultConstructorSourceCode(ConstructorKind cons
     return SourceCode();
 }
 
-UnlinkedFunctionExecutable* BuiltinExecutables::createDefaultConstructor(ConstructorKind constructorKind, const Identifier& name)
+UnlinkedFunctionExecutable* BuiltinExecutables::createDefaultConstructor(ConstructorKind constructorKind, const Identifier& name, bool hasInstanceFields)
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/BuiltinExecutables.cpp]: Ditto.

File: Source/JavaScriptCore/builtins/BuiltinExecutables.cpp [hunk 5]
```diff
@@ -58,30 +58,30 @@ SourceCode BuiltinExecutables::defaultConstructorSourceCode(ConstructorKind cons
...
 {
     switch (constructorKind) {
     case ConstructorKind::None:
         break;
     case ConstructorKind::Base:
     case ConstructorKind::Extends:
-        return createExecutable(m_vm, defaultConstructorSourceCode(constructorKind), name, constructorKind, ConstructAbility::CanConstruct);
+        return createExecutable(m_vm, defaultConstructorSourceCode(constructorKind), name, constructorKind, ConstructAbility::CanConstruct, hasInstanceFields);
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/BuiltinExecutables.cpp]: Ditto.

File: Source/JavaScriptCore/builtins/BuiltinExecutables.cpp [hunk 6]
```diff
@@ -58,30 +58,30 @@ SourceCode BuiltinExecutables::defaultConstructorSourceCode(ConstructorKind cons
...
     }
     ASSERT_NOT_REACHED();
     return nullptr;
 }
 
 UnlinkedFunctionExecutable* BuiltinExecutables::createBuiltinExecutable(const SourceCode& code, const Identifier& name, ConstructAbility constructAbility)
 {
-    return createExecutable(m_vm, code, name, ConstructorKind::None, constructAbility);
+    return createExecutable(m_vm, code, name, ConstructorKind::None, constructAbility, false);
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/BuiltinExecutables.cpp]: Ditto.

File: Source/JavaScriptCore/builtins/BuiltinExecutables.cpp [hunk 7]
```diff
@@ -58,30 +58,30 @@ SourceCode BuiltinExecutables::defaultConstructorSourceCode(ConstructorKind cons
...
 }
 
 UnlinkedFunctionExecutable* createBuiltinExecutable(VM& vm, const SourceCode& code, const Identifier& name, ConstructAbility constructAbility)
 {
-    return BuiltinExecutables::createExecutable(vm, code, name, ConstructorKind::None, constructAbility);
+    return BuiltinExecutables::createExecutable(vm, code, name, ConstructorKind::None, constructAbility, false);
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/BuiltinExecutables.cpp]: Ditto.

File: Source/JavaScriptCore/builtins/BuiltinExecutables.cpp [hunk 8]
```diff
@@ -58,30 +58,30 @@ SourceCode BuiltinExecutables::defaultConstructorSourceCode(ConstructorKind cons
...
 }
 
-UnlinkedFunctionExecutable* BuiltinExecutables::createExecutable(VM& vm, const SourceCode& source, const Identifier& name, ConstructorKind constructorKind, ConstructAbility constructAbility)
+UnlinkedFunctionExecutable* BuiltinExecutables::createExecutable(VM& vm, const SourceCode& source, const Identifier& name, ConstructorKind constructorKind, ConstructAbility constructAbility, bool classContextHasInstanceFields)
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/BuiltinExecutables.cpp]: Ditto.

File: Source/JavaScriptCore/builtins/BuiltinExecutables.h [hunk 9]
```diff
@@ -57,9 +57,9 @@ SourceCode name##Source();
 #undef EXPOSE_BUILTIN_EXECUTABLES
 
     static SourceCode defaultConstructorSourceCode(ConstructorKind);
-    UnlinkedFunctionExecutable* createDefaultConstructor(ConstructorKind, const Identifier& name);
+    UnlinkedFunctionExecutable* createDefaultConstructor(ConstructorKind, const Identifier& name, bool hasInstanceFields);
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/BuiltinExecutables.h]: Ditto.

File: Source/JavaScriptCore/bytecode/ExecutableInfo.h [hunk 10]
```diff
@@ -35,7 +35,7 @@ enum class EvalContextType    : uint8_t { None, FunctionEvalContext, InstanceFie
 // FIXME: These flags, ParserModes and propagation to XXXCodeBlocks should be reorganized.
 // https://bugs.webkit.org/show_bug.cgi?id=151547
 struct ExecutableInfo {
-    ExecutableInfo(bool usesEval, bool isStrictMode, bool isConstructor, bool isBuiltinFunction, ConstructorKind constructorKind, JSParserScriptMode scriptMode, SuperBinding superBinding, SourceParseMode parseMode, DerivedContextType derivedContextType, bool isArrowFunctionContext, bool isClassContext, EvalContextType evalContextType)
+    ExecutableInfo(bool usesEval, bool isStrictMode, bool isConstructor, bool isBuiltinFunction, ConstructorKind constructorKind, JSParserScriptMode scriptMode, SuperBinding superBinding, SourceParseMode parseMode, DerivedContextType derivedContextType, bool classContextHasInstanceFields, bool isArrowFunctionContext, bool isClassContext, EvalContextType evalContextType)
```
Comment (caiolima) [inline Source/JavaScriptCore/bytecode/ExecutableInfo.h]: Ditto.

File: Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp [hunk 11]
```diff
@@ -2831,15 +2831,16 @@ RegisterID* BytecodeGenerator::emitCreateThis(RegisterID* dst)
 
 RegisterID* BytecodeGenerator::emitInstanceFieldInitializationIfNeeded(RegisterID* dst, RegisterID* constructor, const JSTextPosition& divot, const JSTextPosition& divotStart, const JSTextPosition& divotEnd)
 {
-    Ref<Label> done = newLabel();
+    if (!(isConstructor() || isDerivedConstructorContext()) || !classContextHasInstanceFields())
```
Comment (caiolima) [inline Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: Nice!

File: Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp [hunk 12]
```diff
@@ -2831,15 +2831,16 @@ RegisterID* BytecodeGenerator::emitCreateThis(RegisterID* dst)
...
+        return dst;
+
     RefPtr<RegisterID> initializer = emitDirectGetById(newTemporary(), constructor, propertyNames().builtinNames().instanceFieldInitializerPrivateName());
-    emitJumpIfTrue(emitIsUndefined(newTemporary(), initializer.get()), done.get());
+    // ASSERT! IsCallable(initializer)
```
Comment (caiolima) [inline Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: I don't think it is necessary.

File: Source/JavaScriptCore/parser/Nodes.cpp [hunk 13]
```diff
@@ -221,6 +222,7 @@ FunctionMetadataNode::FunctionMetadataNode(
         , m_isInStrictContext(isInStrictContext)
         , m_superBinding(static_cast<unsigned>(superBinding))
         , m_constructorKind(static_cast<unsigned>(constructorKind))
+        , m_classContextHasInstanceFields(false)
```
Comment (caiolima) [inline Source/JavaScriptCore/parser/Nodes.cpp]: Ditto.

File: Source/JavaScriptCore/builtins/BuiltinExecutableCreator.cpp (hunk 3 above)
Comment (caitp) [reply Source/JavaScriptCore/builtins/BuiltinExecutableCreator.cpp]: Done

File: Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp (hunk 12 above)
Comment (caitp) [reply Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp]: Removed

File: Source/JavaScriptCore/runtime/ModuleProgramExecutable.h [hunk 14]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (caiolima) [inline Source/JavaScriptCore/runtime/ModuleProgramExecutable.h]: Oops.

**Comment by xanlpz** [comment]
LGTM. I rebased it manually, fixed a couple of style nits and run the tests again just in case.

**Pull request closed without merging**